Kevinf october updates

kfansler opened the pull request:
A bunch of small commits that I started while waiting for the beta to be final. If there are any Node.js samples in this PR, then theoretically they should be identical to what we just released. I tried to keep them separate.

kfansler committed:
Fixed broken link

APISDK-1212

kfansler committed:
Bug fix

APISDK-1215

kfansler committed:
Fixed admin only ListOrgSheets

C# call pointed to the wrong method.

kfansler committed:
OAuth requires HTTPS

Added notes

kfansler committed:
Rework of the Formatting section

* Added object examples
* Added column of example values to the Format Descriptor table
* Added workflow topic
* Simplified text overall

kfansler committed:
Added sheetVersion

As an include param for
* Get /folders/{folderId}
* Get /sheets
* Get /workspaces/{workspaceId}

kfansler committed:
Clarified exclude flag behavior for conditional formatting

kfansler committed:
Search filters

APISDK-1317, 1319, 1325

kfansler committed:
Clarified contactOptions

APISDK-1217

kfansler committed:
Copy notifications

Clarified include rules to add notifications, deprecate “include=all”

kfansler committed:
Clarified title behavior

kfansler committed:
Fixed Java samples

To new rule: all samples should be able to be pasted into the SDK and
work.
Maps to APISDK-1350

kfansler committed:
Add image to cell

C#, Java, and Python samples.
Node.js sample is in the other branch.

kfansler committed:
Update alt text

Example documented in Update Row and syntax example in cURL.

kfansler committed:
Added groupAdmin

GET /users (already had it)
POST /users
PUT /users

kfansler committed:
New sheet & report publish flags

Corresponds to Proj-0970.

kfansler committed:
List Image URLs issue

Explained the variations a user might see in when error codes appear.

kfansler committed:
Tip on how to use state value

OAuth flow— using state as a reminder value.

seweil commented on source/index.html.md.erb:
```diff
@@ -562,6 +564,8 @@ WRITE_SHEETS | Insert and modify sheet data, including attachments, discussions,
 Your app must implement a 3-legged OAuth flow to retrieve an access token that can be used to access Smartsheet data on behalf of an end user. 
 The following sub-sections describe next steps.
 
+<aside class="notice">App registration and OAuth flow require HTTPS.</aside>
```
(I'm assuming this is in response to the email or support case) As written, this isn't actionable. Needs to be documented as a requirement on the redirect_uri in the token endpoints.

kfansler replied:
Done

seweil commented on source/index.html.md.erb:
```diff
@@ -581,6 +585,8 @@ Here is an example of a URL to send your user as a `GET` request:
 If your user has not logged in to Smartsheet, they are first directed to the Smartsheet login page. 
 After a successful login, Smartsheet prompts your user to allow or deny the access scopes you requested.
 
+<aside class="notice">At the end of this process, your redirect URL must match what you set at the start of this flow. One trick for remembering that or if you have multiple requests is to pass the `state` value to list what redirect URL you plan to use.</aside>
```
Please check with the dev on the email thread. I suspect the restriction is that the final url must be *under* the original. (Otherwise, what's the point?) 
Please remove the 'trick' detail here. Could be part of the 'state' description. perhaps: "State may also be used to distinguish between multiple authorization flows."

kfansler replied:
I reworked it. Mostly using the text you suggested. But I additionally reworded your text as an aside. It's probably overkill.

seweil commented on source/index.html.md.erb:
```diff
@@ -2795,15 +2801,46 @@ curl https://api.smartsheet.com/2.0/sheets/{sheetId}/rows/{rowId}/columns/{colum
+// Set options
+Image image = new Image();
+image.setAltText("Caution sign");
+image.setHeight(16L);
```
Are these really long integers?

kfansler replied:
Yes. I just re-tried it and the SDK calls for a Long, and then you get the red squiggles underneath if you don't include the L indicator.

seweil commented on source/index.html.md.erb:
```diff
@@ -3019,7 +3056,7 @@ String temporaryUrl = ss.imageUrlResources().getImageUrls(imageUrls).getImageUrl
 			"error": {
 				"errorCode": 5001,
-				"message": "localized_error_message"
+				"message": "customizable error message"
```
"Customizable" implies that the client can change it. These are errors that we return to them. If you can get a concrete example from Daniel, that would be great. Else something like "You do not have permission to view this image."

kfansler replied:
Hmm. Here's what our error code table shows: "400 | 5xxx | Errors in the 5xxx range represent conditions that a developer cannot reasonably prevent or handle, most typically related to account status. These error messages are localized and can be displayed to the end-user to inform them of the condition that caused the error to occur." 
My personal take is that "localized" is being used incorrectly in both places. I think we should fix it in both places once we agree on a term.  "reserved" or the example you give works for the actual sample, and then some other substitution for the table?

seweil commented on source/index.html.md.erb:
```diff
@@ -3030,11 +3067,7 @@ String temporaryUrl = ss.imageUrlResources().getImageUrls(imageUrls).getImageUrl
-<aside class="notice">The number and sequence of <a href="#imageurl-object">ImageUrl Objects</a> in the Response corresponds exactly to the number and sequence of objects specified by the Request.</aside>
-
-<aside class="notice">If an error is encountered for any of the requested images, the corresponding <b>ImageUrl</b> Object in the Response contains an <b>error</b> attribute that describes the error. (See Example Response.)</aside>
+<aside class="notice">If an error is encountered for any of the requested images, the corresponding <b>ImageUrl</b> Object in the Response contains an <b>error</b> attribute that describes the error. (See Example Response.) In addition, some errors only appear when you try to reference the URL.</aside>
```
I always try to make things as concrete as possible. How about "Additional error checking is performed when the returned URL is referenced." or "Certain errors are not detected until the returned URL is referenced."

kfansler replied:
Done.

seweil commented on source/index.html.md.erb:
```diff
@@ -6430,7 +6464,7 @@ Gets the specified Folder (and lists its contents).
 <%= SCOPE %> | READ_SHEETS
 <%= HEADERS %> | <%= HEADER_AUTH %>
-<%= PARAMETERS %> | **include** (optional) -- a comma-separated list of optional elements to include in the response:<ul><li>ownerInfo - owner's email address and user Id for each sheet</li><li>source - the [source](#source-object) for any sheet that was created from another sheet or template</li></ul>
+<%= PARAMETERS %> | **include** (optional) -- a comma-separated list of optional elements to include in the response:<ul><li>ownerInfo - owner's email address and user Id for each sheet</li><li>sheetVersion - current version number of the sheet, should not be combined with pagination</li><li>source - the [source](#source-object) for any sheet that was created from another sheet or template</li></ ...
```
current version of *each* sheet

kfansler replied:
Done

seweil commented on source/index.html.md.erb:
```diff
@@ -7680,9 +7714,11 @@ Extends the [Row Object](#row-object), adding the following:
+**readOnlyFullAccessibleBy** | string | Flag to indicate who can access the 'Read-Only Full' view of the published report:<ul><li>If "ALL", it is available to anyone who has the link.</li><li>If "ORG", it is available only to members of the report owner's Smartsheet organization.</li></ul>Only returned in a response if **readOnlyFullEnabled** = true.
+**readOnlyFullDefaultView** | enum | Indicates which view the user has set for a read-only, default view of the published report. 0 or null = GRID, 1 = CARD, and 2 = CALENDAR.
 **readOnlyFullEnabled** | Boolean | If `true`, a rich version of the report is published with the ability to download row attachments and discussions.
+**readOnlyFullShowToolbar** | Boolean | If `true`, shows the left nav toolbar.
```
True is the default. So if false - it is hidden

kfansler replied:
I clarified it.

seweil commented on source/index.html.md.erb:
```diff
@@ -9407,7 +9443,13 @@ Searches all Sheets that the User can access, for the specified text.
 <%= HEADERS %> | <%= HEADER_AUTH %>
-<%= PARAMETERS %> | **query** (required): Text with which to perform the search. Enclose in double-quotes for an exact search.
+<%= PARAMETERS %> | **query** (required): text with which to perform the search. Enclose in double-quotes for an exact search.
+ | **include** (optional): when specified with a value of "favoriteFlag", response includes:<ul><li>favorite -- Folders, Reports, Sheets, Sights, Templates, and Workspaces will have the property "favorite": true</li><li>parentObjectFavorite -- Attachments, Discussions, Profile Fields, and Rows will have the property "parentObjectFavorite": true</li></ul>
```
... response indicates which returned items are favorites

kfansler replied:
done

seweil commented on source/index.html.md.erb:
```diff
@@ -9407,7 +9443,13 @@ Searches all Sheets that the User can access, for the specified text.
-<%= PARAMETERS %> | **query** (required): Text with which to perform the search. Enclose in double-quotes for an exact search.
+<%= PARAMETERS %> | **query** (required): text with which to perform the search. Enclose in double-quotes for an exact search.
+ | **include** (optional): when specified with a value of "favoriteFlag", response includes:<ul><li>favorite -- Folders, Reports, Sheets, Sights, Templates, and Workspaces will have the property "favorite": true</li><li>parentObjectFavorite -- Attachments, Discussions, Profile Fields, and Rows will have the property "parentObjectFavorite": true</li></ul>
+ | **level** (optional): when specified with one of the following values, temporarily disables filters:<ul><li>0 -- allows all filters except for Dashboard, Folder, or Workspace</li><li>1 -- allows all filters except for Folder or Workspace</li><li>2 -- allows all filters</li></ul>
```
level flags are internal (undocumented) for mobile. Please remove

kfansler replied:
Done.

seweil commented on source/index.html.md.erb:
```diff
@@ -9407,7 +9443,13 @@ Searches all Sheets that the User can access, for the specified text.
+ | **level** (optional): when specified with one of the following values, temporarily disables filters:<ul><li>0 -- allows all filters except for Dashboard, Folder, or Workspace</li><li>1 -- allows all filters except for Folder or Workspace</li><li>2 -- allows all filters</li></ul>
+ | **location** (optional): when specified with a value of "personalWorkspace", limits the response to only those items in the user's Workspaces.
+ | <%= PARAMETER_MODIFIEDSINCE %>
+ | **owner** (optional): uses a primary email address, such as "Jane.Roe@smartsheet.com", to filter responses to only those items owned by that user.
```
owner is not implemented

kfansler replied:
Done

kfansler committed:
Incorporated review comments

seweil commented on source/index.html.md.erb:
```diff
@@ -11125,16 +11167,21 @@ A Sheet can exist in a user's **Sheets** folder ([Home](#home)), in a [Folder](#
+**readOnlyFullShowToolbar** | Boolean | If `true`, shows the left nav toolbar.
 **readOnlyFullUrl** | string | URL for 'Read-Only Full' view of the published sheet<br/>Only returned in a response if **readOnlyFullEnabled** = true.
+**readOnlyLiteEnabled** | Boolean | If `true`, a lightweight version of the sheet is published without row attachments or discussions.
+**readOnlyLiteSSslUrl** | string | If `true`, returns an SSL link.
```
This is an URL - never 'true'. Returned if the the sheet is published to an SSL url

kfansler replied:
Yeah, you're right. I don't see how a string can also be a Boolean. The spec specifically mentions being able to set it to false so that nothing will be returned, but it also says it's a string. So either it's a Boolean or it's a string. It's Proj-0970 if you want to help me decipher/interpret.

seweil commented on source/index.html.md.erb:
```diff
@@ -11125,16 +11167,21 @@ A Sheet can exist in a user's **Sheets** folder ([Home](#home)), in a [Folder](#
+**readOnlyLiteEnabled** | Boolean | If `true`, a lightweight version of the sheet is published without row attachments or discussions.
+**readOnlyLiteSSslUrl** | string | If `true`, returns an SSL link.
+**readOnlyLiteUrl** | string | URL for 'Read-Only HTML' view of the published sheet<br/>Only returned in a response if **readOnlyLiteEnabled** = true.
+**readWriteAccessibleBy** | string | Flag to indicate who can access the 'Edit by Anyone' view of the published sheet:<ul><li>If "ALL", it is available to anyone who has the link.</li><li>If "ORG", it is available only to members of the sheet owner's Smartsheet organization.</li></ul>Only returned in a response if **readWriteEnabled** = true.
```
An enum (can be one of several values)
Let's avoid the term 'flag'

kfansler replied:
I can rewrite the descriptions that use Flag; however, that's not really new for this PR. I just alphabetized the list, so you're seeing a change.

seweil commented on source/index.html.md.erb:
```diff
@@ -15852,46 +15903,81 @@ HTTP status code | Smartsheet errorCode | Smartsheet message
 # Formatting
 
-Formatting data can optionally be included for columns, rows, and cells by adding the "include=format" query string parameter to any API operation that returns any of those objects (for example, `GET /sheets/{sheetId}`, `GET sheets/{sheetId}/rows/{rowId}`, etc.).  When this parameter is included, objects that contain non-default formatting include a **format** property.  Objects that do not have  ...
+You can format data in cells, columns, or rows by using the optional "include=format" query-string parameter with API operations that return any of those objects (for example, `GET /sheets/{sheetId}` or `GET sheets/{sheetId}/rows/{rowId}`). When you include this parameter, and if there is formatting other than default settings, the return includes a **format** property.
```
This query param determines what is *returned*, it does not format cells, etc. It allows you to *read* the current formatting

kfansler replied:
Clarified this part.

seweil commented on source/index.html.md.erb:
```diff
@@ -15852,46 +15903,81 @@ HTTP status code | Smartsheet errorCode | Smartsheet message
+
+1. Use `GET /serverinfo` to retrieve the predefined format lookup table.
+2. Look through the [FormatTables Object](#formattables-object) to map the format property to the value you need, for instance, fontFamily is the first position. If you want to change the font to Tahoma, you need to make a note that Tahoma is the 2nd value for fontFamily, so it has a value of 1 (the default or first value is 0).
+3. Go to the [Format Descriptor](#format-descriptor) section for quick reference.
```
Needs more discussion - let's talk

kfansler committed:
Most review comments incorporated

Still outstanding is a discussion on formatting.

kfansler committed:
Deprecated the redirect-url value

kfansler committed:
Removed alternate usage for state

kfansler committed:
Rewrite of formatting section

I streamlined this and gave better examples. I think the flow of
information is better and more in line with how other sections in the
ref are formatted.

seweil commented on source/index.html.md.erb:
```diff
@@ -13832,7 +13876,7 @@ Gets a list of user-created Templates that the user has access to.
 # Token
 
-The Smartsheet API utilizes OAuth 2.0 for Authentication and Authorization. An **Authorization** HTTP header containing an Access Token is required to authenticate requests.  For more information, see [Authentication](#authentication-and-access-tokens).
+The Smartsheet API utilizes OAuth 2.0 for Authentication and Authorization. An **Authorization** HTTPS header containing an Access Token is required to authenticate requests.  For more information, see [Authentication](#authentication-and-access-tokens).
```
AFAICT, the convention is to call these HTTP headers, even if the actual transport is HTTPS. (Several places)

kfansler replied:
Done. I only found two places.

seweil commented on source/index.html.md.erb:
````diff
@@ -15852,48 +15897,70 @@ HTTP status code | Smartsheet errorCode | Smartsheet message
+> Example defaults value:
 
-## Row & Column Format
+```json
````
This feels out of place in the right column - no anchored to any object

kfansler replied:
I had trouble finding this--not sure if I found the right thing. The line numbers don't match up with the MD file line numbers. And a search got me nowhere. I *think* you were objecting to the json sample I had of the defaults value that was out of context, so I've removed that.

seweil commented on source/index.html.md.erb:
```diff
@@ -15852,48 +15897,70 @@ HTTP status code | Smartsheet errorCode | Smartsheet message
-The format descriptor contains 16 comma-separated numeric indexes.  Each index represents a format style whose value can be found in one of the lookup tables in the [FormatTables Object](#formattables-object).
+**Position** | **Lookup Property** | **Example Value** | **Model**
+-----|-----|-----|-----|
+0 | fontFamily | 0 = Arial, default | "0,,,,,,,,,,,,,,,"
```
I like this table.
Can we get the samples in a monospaced (code) font? 
I think the Examples should match the Models
Rather than 'Model' - how about 'format string'?

kfansler replied:
Done.

seweil commented on source/index.html.md.erb:
```diff
@@ -15913,20 +15980,63 @@ The FormatTables Object is retrieved via the `GET /serverinfo` operation and con
+
+## Set Formatting for a Row or Column
+
+Setting the format of a row or column through the API simply sets the baseline format for new cells in that row or column.  It does not affect existing cells.
```
Have you tested setting formatting for a row or col? (I'm unclear on what's possible)

kfansler replied:
Yes. It won't change any cells that already have values, but if you change either a row or column to italic (for instance), then any newly inputted text will be italic. Even a picklist will type in as italic, and then revert to the picklist formatting when you hit enter.

seweil commented on source/index.html.md.erb:
```diff
@@ -15913,20 +15980,63 @@ The FormatTables Object is retrieved via the `GET /serverinfo` operation and con
+ 
+To do something similar to highlighting an entire row or column in the UI and changing the format to bold, use one of the following API calls:
+
+* `PUT /sheets/{sheetId}/rows/{rowId}`
```
The user needs to set format on each cell

kfansler replied:
Not that I can tell from testing. You can do an entire row or column. Perhaps in the SDKs you have to create a loop? Not sure on the variations. But in Postman, you can do an entire row or column in one go.

kfansler committed:
More review comments incorporated.

kfansler committed:
Incorp final discussion on formatting

Clarify how to format cells with data vs rows or column cells that have
no data.

kfansler committed:
wordsmithing

seweil committed:
Merge branch 'master' into kevinf-october-updates
